### ISSUE TITLE
Heap size for sbcl

### ISSUE THREAD

**Issue by Gillani0** [aaltodsg/instans#34]
Hi, 
So I am getting an error if I load a file of a medium size (125 MB, around 200K triples).
Is it any way to read the file triple by triple rather than loading it once?

Following is the error that I received. Cheers

Heap exhausted during garbage collection: 112 bytes available, 224 requested.
 Gen StaPg UbSta LaSta LUbSt Boxed Unboxed LB   LUB  !move  Alloc  Waste   Trig    WP  GCs Mem-age
   0:     0     0     0     0     0     0     0     0     0        0     0 10737418    0   0  0.0000
   1:     0     0     0     0     0     0     0     0     0        0     0 10737418    0   0  0.0000
   2: 31196 31352     0     0  2379  7071     0     0     0 308895424 762176 195709466    0   1  1.4059
   3: 31955 32767     0     0  6458 13014   305   471    26 660975008 2511456  2000000 4223   0  1.2563
   4:     0     0     0     0     0     0     0     0     0        0     0  2000000    0   0  0.0000
   5:     0     0     0     0     0     0     0     0     0        0     0  2000000    0   0  0.0000
   6:     0     0     0     0  2307   763     0     0     0 100597760     0  2000000 2059   0  0.0000
   Total bytes allocated    = 1070468192
   Dynamic-space-size bytes = 1073741824
GC control variables:
   *GC-INHIBIT* = true
   *GC-PENDING* = true
   *STOP-FOR-GC-PENDING* = false
fatal error encountered in SBCL pid 8025(tid 140737354057472):
Heap exhausted, game over.

**Comment by MJRinne** [comment]
There are quite many answers to that question:

1) There is a problem with memory at the moment, I noticed it last Sunday. Independent of what is done, stuff keeps accumulating in memory. Needs to be fixed.

2) INSTANS doesn't natively clear anything from memory, this needs attention from the user. Basically you need to either [DELETE the data, which is no longer needed](https://github.com/aaltodsg/instans/wiki/Memory-Management) or use the [remove input-policy](https://github.com/aaltodsg/instans/wiki/RDF-operations) to get rid of incoming events, which have already been processed.

3) As INSTANS is intended to be an event stream processor, it is doing continuous processing (which is why "loading a file" sounds a bit biased :-). To print the answers on-the-fly, you need to use the "flush" [operation policy](https://github.com/aaltodsg/instans/wiki/RDF-operations).

4) Depending on the [input file type](https://github.com/aaltodsg/instans/wiki/RDF-file-formats), INSTANS is processing the data either triple-by-triple or in blocks. Loading the whole document is never the default. The --rdf-input-unit can be selected with a [command line parameter](https://github.com/aaltodsg/instans/wiki/Command-line-parameters). I normally use "block", because I'd like my events to be processed as complete entities.

So, looking at your question, what I would do now is:

A) Add an "--rdf-operations=add:execute:remove:flush" parameter on the command line after the queries but before the data. The "remove" should clean the input events after all related queries have been processed and "flush" should output the answers on the fly, which would allow you to see whether the output is as expected. If the query network is more convoluted, alternatively "add:execute:remove:execute:flush" may be needed.

B) Add an "--rdf-input-unit=triple" command line parameter before the data file, if your data is indeed unstructured triples.

This won't process your entire stream before the real problem 1) is fixed, but hopefully allows you to at least verify the output until the point where the heap runs out.

Cheers,

**Comment by Gillani0** [comment]
Hi, thanks for the detailed response.

But I am confused about couple of things.
i) In block input mode, how the system is filtering or waiting for a set of triples. For instance if I have a N-Quad file (I know the preferred input should be in Tigx format) with the following triples, and if I use the block input mode, will it process them according to the Quads (or Named graph) IDs?
Because I am using the block wise input and if the data of each event exceeds a certain limit (e.g., 100K triples), I still gets the out of memory error. Cheers

## First Event ###
<http://www.ldbc.eu/ldbc_socialnet/1.0/data/pers00000004398046518943> <http://www.ldbc.eu/ldbc_socialnet/1.0/vocabulary/creationDate> "1981-08-17T07:33:47Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> <http://example.org/event/1> .

<http://www.ldbc.eu/ldbc_socialnet/1.0/data/pers00000004398046518943> <http://www.ldbc.eu/ldbc_socialnet/1.0/vocabulary/isLocatedIn> <http://dbpedia.org/resource/Panama_City> <http://example.org/event/1> .

<http://www.ldbc.eu/ldbc_socialnet/1.0/data/pers00000004398046518943> <http://www.ldbc.eu/ldbc_socialnet/1.0/vocabulary/speaks> "es" <http://example.org/event/1> .

####Second Event####
<http://www.ldbc.eu/ldbc_socialnet/1.0/data/pers00000004398046518943> <http://www.ldbc.eu/ldbc_socialnet/1.0/vocabulary/creationDate> "1981-08-17T07:33:47Z"^^<http://www.w3.org/2001/XMLSchema#dateTime> <http://example.org/event/2> .

<http://www.ldbc.eu/ldbc_socialnet/1.0/data/pers00000004398046518943> <http://www.ldbc.eu/ldbc_socialnet/1.0/vocabulary/isLocatedIn> <http://dbpedia.org/resource/Panama_City> <http://example.org/event/2> .

<http://www.ldbc.eu/ldbc_socialnet/1.0/data/pers00000004398046518943> <http://www.ldbc.eu/ldbc_socialnet/1.0/vocabulary/speaks> "es" <http://example.org/event/2> .

**Comment by MJRinne** [comment]
Nope, for [NQuads and Ntriples block = single](https://github.com/aaltodsg/instans/wiki/RDF-file-formats). Based on http://www.w3.org/TR/2014/REC-n-quads-20140225/#sec-grammar and http://www.w3.org/TR/2014/REC-n-triples-20140225/#n-triples-grammar there is no possibility to form larger entities, whereas with [TriG](http://www.w3.org/TR/2014/REC-trig-20140225/#grammar-ebnf) the block is bigger.

You have fairly large events; we have typically had only event sizes up to some tens of triples. If they come in an NQuad-file they are processed as individual quads. 100K events could be hit by the current memory issue, and the situation could also improve just by increasing the maximum heap size, which is currently at 1G.

**Comment by MJRinne** [comment]
Some memory leakage problems still remain, but looks like the main problem is the TriG parser, so the situation may actually be better using nquads. With the new build process the maximum heap size is user-selectable. After running `./configure` to create a makefile, the makefile will contain the following variable:

    DYNAMIC_SPACE_SIZE=4096

The value is in megs, so the default is now 4G (earlier default heap max was 1G). If there are still problems with heap size, larger values can be tried.
